src/Features/LanguageServer/Protocol/RoslynRequestExecutionQueue.cs

```diff
@@ -39,6 +44,48 @@ public override Task WrapStartRequestTaskAsync(Task nonMutatingRequestTask, bool
             }
         }
 
+        /// <inheritdoc/>
+        protected override IMethodHandler GetHandlerForRequest(IQueueItem<RequestContext> work)
+        {
+            var handlers = _handlerProvider.GetMethodHandlers(work.MethodName, work.RequestType, work.ResponseType);
+            var defaultHandler = handlers.First(h => string.IsNullOrEmpty(h.Metadata)).Value;
+            var identifier = RoslynRequestExecutionQueue.GetTextDocumentIdentifier(work, defaultHandler);
+            if (identifier is null)
+            {
+                return defaultHandler;
+            }
+
+            var language = _lspWorkspaceManager.GetLanguageForUri(identifier.Uri);
+
+            return handlers.SingleOrDefault(h => string.Equals(h.Metadata, language, StringComparison.OrdinalIgnoreCase))?.Value ?? defaultHandler;
+        }
+
+        protected static TextDocumentIdentifier? GetTextDocumentIdentifier(IQueueItem<RequestContext> work, IMethodHandler handler)
+        {
+            var textIdentifier = work.GetTextDocumentIdentifier<TextDocumentIdentifier>(handler);
+            if (textIdentifier != null)
+            {
+                return textIdentifier;
+            }
+
+            var nullIdentifier = work.GetTextDocumentIdentifier<TextDocumentIdentifier?>(handler);
```

Comment:
i don't undersatand how this line is different from line 65.

Reply:
This accounts for different handler signatures. Some have the TextDocumentIdentifier nullable and some don't.

Reply:
but it's a class.  So it makes no difference.

Reply:
I think the IQueueItem is generic - we should probably constrain the generic definition there to class.